### PR TITLE
Manual grading: random assignment of graders to questions

### LINKED ISSUE #9327: random assignment of graders to assessment instances
It would be helpful if PrairieLearn could automatically assign all instances of a manually graded question to a set of pre-defined graders. This would be a random grader assignment so that every grader gets a similar number of instances to grade. 

This will be helpful when the class is very large, and questions are very easy to grade using the rubric, so you don't need to assign all instances to the same grader for consistency (which can be easily done with the current interface). If I want to split the work among a given number of grader, I need to manually click a bunch of boxes, making sure I have the desired number of boxes selected, and then assign to a grader. Not super awful, but it would be much better to do that by clicking just a couple of buttons (select group of graders, click select all assessments, then click randomly assign :-))

### PR DESCRIPTION
**Pull request by jonatanschroeder**
Resolves #9327. Also converts the sproc course_instance_select_graders into a model function.

### CODE REVIEW

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.ts [hunk 1]
```diff
@@ -24,7 +34,62 @@ router.get(
...
+    }
+    if (req.body.__action === 'assign_graders') {
+      if (!req.body.assigned_grader) {
+        flash('error', 'No graders were selected for assignment.');
+        res.redirect(req.originalUrl);
+        return;
+      }
+      const graderIds: string[] = Array.isArray(req.body.assigned_grader)
+        ? req.body.assigned_grader
+        : [req.body.assigned_grader];
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.ts:65]: AFAICT we're completely lacking any validation that these users are student data editors within this course. Did I miss that somewhere?

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts [hunk 2]
```diff
@@ -51,11 +53,9 @@ async function prepareLocalsForRender(query: Record<string, any>, resLocals: Rec
     }
   }
 
-  const graders = await sqldb.queryOptionalRow(
-    sql.select_graders,
-    { course_instance_id: resLocals.course_instance.id },
-    UserSchema.array().nullable(),
-  );
+  const graders = await selectUsersWithCourseInstanceAccess({
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts:56]: That's the default:
https://github.com/PrairieLearn/PrairieLearn/blob/8094110c978fa78ec8424d208599002c70901690/apps/prairielearn/src/models/course-instances.ts#L77

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts [hunk 3]
```diff
@@ -172,3 +189,47 @@ function ProgressBar(partial: number | null, total: number | null) {
...
+    id: 'grader-assignment-modal',
+    title: 'Assign instances to graders',
+    body: html`
+      <p>Assign instances to the following graders:</p>
+      ${courseStaff.map(
+        (staff) => html`
+          <div class="form-check">
+            <input
+              type="checkbox"
+              id="graderAssignment${staff.user_id}"
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts]: Done in 31807dbc4.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.ts (hunk 1 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.ts]: Indeed it was missing. Done in 31807dbc4.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts:56]: That feels like a footgun to me? It's not at all obvious that that would or should be the default. If you're introducing this new API, I'd rather require the caller to always be explicit about what type of "course instance access" they're querying for.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts [hunk 4]
```diff
@@ -172,3 +189,47 @@ function ProgressBar(partial: number | null, total: number | null) {
...
+        If more than one grader is selected, the instances will be randomly split between the
+        graders.
+      </div>
+    `,
+    footer: html`
+      <input type="hidden" name="unsafe_assessment_question_id" value="" />
+      <input type="hidden" name="__csrf_token" value="${csrfToken}" />
+      <input type="hidden" name="__action" value="assign_graders" />
+      <button type="submit" class="btn btn-primary">Assign</button>
+      <button type="button" class="btn btn-secondary" data-dismiss="modal">Cancel</button>
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:232]: Was this copied from somewhere? Typically we'd want the cancel/dismiss button on the left and the affirmative action button on the right.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts [hunk 5]
```diff
@@ -172,3 +189,47 @@ function ProgressBar(partial: number | null, total: number | null) {
...
+              value="${staff.user_id}"
+              class="form-check-input"
+            />
+            <label class="form-check-label" for="grader-assignment-${staff.user_id}">
+              ${staff.name ? `${staff.name} (${staff.uid})` : staff.uid}
+            </label>
+          </div>
+        `,
+      )}
+      <div class="mt-2 small alert alert-info">
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:221]: Nit, improved spacing:

```suggestion
      <div class="mt-3 mb-0 small alert alert-info">
```

**Review by nwalters512 — COMMENTED** [review]
A few more comments. I'd like to finish discussion on https://github.com/PrairieLearn/PrairieLearn/pull/10305/files#r1704471969 before approving.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:232]: I think this was copilot 😬 ... I'll change it.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts]: I'd prefer not to have the explicit permission required in the call itself, mostly to abstract the "grader" role with the permission it requires. In 781e26415ef7b2a5ff1ebfde3f78ce7413642247 I made the role required, but I added a helper that calls this function with the proper permission.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:232]: I just double checked, and some other modals in manual grading also use this order: [rubric settings](https://github.com/PrairieLearn/PrairieLearn/blob/781e26415ef7b2a5ff1ebfde3f78ce7413642247/apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/rubricSettingsModal.html.ts#L260) and [grading conflict info on assessment question](https://github.com/PrairieLearn/PrairieLearn/blob/781e26415ef7b2a5ff1ebfde3f78ce7413642247/apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessmentQuestion/assessmentQuestion.html.ts#L132). Should I change those too?

Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts]: Yes, though maybe in another PR that we can land independently?

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts]: Moved to #10414.

File: apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentClient.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+      const assessmentQuestionInput = document.querySelector<HTMLInputElement>(
+        '#grader-assignment-modal input[name="unsafe_assessment_question_id"]',
+      );
+      if (assessmentQuestionInput) {
+        assessmentQuestionInput.value = button.dataset.assessmentQuestionId ?? '';
+      }
+      $('#grader-assignment-modal').modal('show');
+    }),
+  );
+});
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentClient.ts:15]: Nit, consider writing this with [`templateFromAttributes`](https://github.com/PrairieLearn/PrairieLearn/blob/master/packages/browser-utils/README.md#templatefromattributes) (we use this everywhere in PT where we need to template a modal, and I'd like to try to do the same in PL):

```suggestion
onDocumentReady(() => {
  const modal = document.getElementById('#grader-assignment-modal');
  document.querySelectorAll<HTMLButtonElement>('.js-assign-grader-btn').forEach((button) =>
    button.addEventListener('click', () => {
      templateFromAttributes(button, modal, {
        'data-assessment-question-id': 'input[name="unsafe_assessment_question_id"]',
      });
      $('#grader-assignment-modal').modal('show');
    }),
  );
});
```

Also consider opening the modal with the data attribute API (`data-toggle`, `data-target`, and such).

Ideally we'd do event handling with `delegated-events`, but we need BS5 to be universally used first.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:1]: Nit, you could use some `align-middle` classes on table cells to improve vertical alignment when the assignment button is shown:

Screenshot 2024-08-15 at 10 27 21

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts [hunk 7]
```diff
@@ -188,3 +205,47 @@ function ProgressBar(partial: number | null, total: number | null) {
...
+              name="assigned_grader"
+              value="${staff.user_id}"
+              class="form-check-input"
+            />
+            <label class="form-check-label" for="grader-assignment-${staff.user_id}">
+              ${staff.name ? `${staff.name} (${staff.uid})` : staff.uid}
+            </label>
+          </div>
+        `,
+      )}
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:236]: Nit, consider adding an empty state when there are no grader users:

Screenshot 2024-08-15 at 10 40 09

**Review by nwalters512 — APPROVED** [review]

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts (hunk 7 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts]: Done in 72dba21b71bb681b019777070e4c735e9b8711be.

File: apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorAssessmentManualGrading/assessment/assessment.html.ts:1]: Done in 884223d9a6c85750677f0396d78e8faf598d3d2d.

File: apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentClient.ts (hunk 6 above)
Comment (jonatanschroeder) [reply apps/prairielearn/assets/scripts/instructorAssessmentManualGradingAssessmentClient.ts]: Switched to data-toggle in af59831b8c26ae32d42bbf34fb2b5659fbdf55cc.

**Pull request merged**